src/Interfaces.cs

```diff
@@ -103,5 +107,29 @@ public static Task<T> DispatchAsync<T>(this IDispatchQueue queue, Func<T> func)
             });
             return tcs.Task;
         }
+
+        /// <summary>
+        /// This allows you await directly on a queue, which is handy for queue-jumping if you are already in the middle of an async method.
+        /// </summary>
+        /// <param name="queue">The queue to jump to with your await statement</param>
+        public static DispatchQueueAwaiter GetAwaiter(this IDispatchQueue queue) => new DispatchQueueAwaiter(queue);
+
+        public struct DispatchQueueAwaiter : INotifyCompletion
+        {
+            readonly IDispatchQueue m_queue;
+
+            public DispatchQueueAwaiter(IDispatchQueue queue)
+            {
+                m_queue = queue;
+                IsCompleted = false;
+            }
+
+            public bool IsCompleted { get; }
+
+            public void OnCompleted(Action continuation) => m_queue.DispatchAsync(continuation); // can't cancel here
```

Comment:
Is this always going to switch to the thread pool? Can you avoid this in some cases - e.g. if you're already on the queue?

Reply:
yeah it will always switch to the threadpool. This is deliberate because while I agree that it would be a nice optimisation to complete without any async things, checking if we're already on the queue relies on the ThreadLocal stack (and associated performance hit) which I want to allow people to opt out of in future